src/ql/pass/map/qubits/place_mip/detail/impl.cc

```diff
@@ -0,0 +1,453 @@
+/** \file
+ * Initial placement engine.
+ */
+
+#include "impl.h"
+
+#include <thread>
+#include <mutex>
+#include <condition_variable>
+#include "Highs.h"
+#include "io/FilereaderMps.h"
+#include "util/HighsMatrixPic.h"
+
+namespace ql {
+namespace pass {
+namespace map {
+namespace qubits {
+namespace place_mip {
+namespace detail {
+
+namespace {    
+static constexpr double EPSILON = 0.000000001;
+
+bool isPermutation(const utils::Vec<utils::UInt> &v2r) {
+    std::set<utils::UInt> image;
+
+    for(auto x: v2r) {
+        if (x < 0 || x >= v2r.size() || image.count(x) > 0) {
+            return false;
+        }
+        image.insert(x);
+    }
+
+    if (image.size() != v2r.size()) {
+        return false;
+    }
+
+    return true;
+}
+
+void logNewMapping(const utils::Vec<utils::UInt> &v2r) {
+    std::stringstream ss;
+    ss << "Initial placement resulted in the following mapping (virtual -> real): { ";
+    bool first = true;
+    for(utils::UInt i = 0; i < v2r.size(); ++i) {
+        if (!first) {
+            ss << ", ";
+        }
+        first = false;
+        ss << i << " -> " << v2r[i];
+    }
+    ss << " }";
+    QL_IOUT(ss.str());
+}
+}
+
+using namespace utils;
+
+/**
+ * String conversion for initial placement results.
+ */
+std::ostream &operator<<(std::ostream &os, Result ipr) {
+    switch (ipr) {
+        case Result::ANY:       os << "any";        break;
+        case Result::CURRENT:   os << "current";    break;
+        case Result::NEW_MAP:   os << "newmap";     break;
+        case Result::FAILED:    os << "failed";     break;
+        case Result::TIMED_OUT: os << "timedout";   break;
+    }
+    return os;
+}
+
+bool Impl::hasNonNN2QGates() {
+    QL_ASSERT(!twoQGatesCount.empty());
+
+    for (const auto& kv: twoQGatesCount) {
+        QL_ASSERT(kv.first.first != kv.first.second);
+        if (distanceProvider(kv.first.first, kv.first.second) > 1) {
+            return true;
+        }
+    }
+
+    return false;
+}
+
+Vec<Vec<UInt>> Impl::computeCostMax(const Vec<Vec<UInt>> &refcount) {
+    Vec<Vec<UInt>>  costmax;
+    costmax.resize(nfac);
+    for (UInt i=0; i<nfac; i++) {
+        costmax[i].resize(qubitsCount, 0);
+    }
+
+    for (UInt i = 0; i < nfac; i++) {
+        for (UInt k = 0; k < qubitsCount; k++) {
+            for (UInt j = 0; j < nfac; j++) {
+                for (UInt l = 0; l < qubitsCount; l++) {
+                    QL_ASSERT(distanceProvider(k, l) < utils::MAX && "All qubits in the topology should be connected");
+                    costmax[i][k] += refcount[i][j] * distanceProvider(k, l);
+                }
+            }
+        }
+    }
+
+    return costmax;
+}
+
+std::unique_ptr<HighsModel> Impl::createHiGHSModel(const Vec<Vec<UInt>> &refcount) {
+    // This describes the MIP problem using HiGHS's API. Some examples can be found in deps/highs/examples.
+
```

Comment:
What I miss is the MIP problem formulation and a very clear description how it is translated to the API as was in the old implementation.

Reply:
In this specific case it is worth it to add a comment indeed. This API is not the simplest... WIll do.

On the side, I just want to point out that in a lot of cases there are also drawbacks to having too many comments:
- comments and code needs to be kept in sync, and in my experience this is hard and tiresome
- comments are not needed when the code is sufficiently clear
- comments are one more (unchecked by the compiler) source of typos, errors, confusion... In the end they can make the code hard to read, even though their goal is the opposite.

In OpenQL, I disagree that there needs to be a comment for every function:
-  A function named "computeCostmax" does not need to have an accompanying comment "This function computes costmax"
- Currently there are comments for both declarations in the headers and in the implementation. Those comments are supposed to be identical I think, but it's also a lot of useless text...
I know there is some doxygen documentation but it's also fine to leave some comments out.